// @suprsend/node-sdk/dist/brands.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;
var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _signature = _interopRequireDefault(require("./signature"));
var _utils = require("./utils");
var _axios = _interopRequireDefault(require("axios"));
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { (0, _defineProperty2["default"])(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
var BrandsApi = /*#__PURE__*/function () {
  function BrandsApi(config) {
    (0, _classCallCheck2["default"])(this, BrandsApi);
    this.config = config;
    this.list_url = this.__list_url();
    this.__headers = this.__common_headers();
  }
  (0, _createClass2["default"])(BrandsApi, [{
    key: "__list_url",
    value: function __list_url() {
      var list_uri_template = "".concat(this.config.base_url, "v1/brand/");
      return list_uri_template;
    }
  }, {
    key: "__common_headers",
    value: function __common_headers() {
      return {
        "Content-Type": "application/json; charset=utf-8",
        "User-Agent": this.config.user_agent
      };
    }
  }, {
    key: "cleaned_limit_offset",
    value: function cleaned_limit_offset(limit, offset) {
      var cleaned_limit = typeof limit === "number" && limit > 0 && limit <= 1000 ? limit : 20;
      var cleaned_offset = typeof offset === "number" && offset >= 0 ? offset : 0;
      return [cleaned_limit, cleaned_offset];
    }
  }, {
    key: "__dynamic_headers",
    value: function __dynamic_headers() {
      return {
        Date: new Date().toUTCString()
      };
    }
  }, {
    key: "list",
    value: function () {
      var _list = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee() {
        var kwargs,
          limit,
          offset,
          _this$cleaned_limit_o,
          _this$cleaned_limit_o2,
          cleaned_limit,
          cleaner_offset,
          final_url_obj,
          final_url_string,
          headers,
          signature,
          response,
          _args = arguments;
        return _regenerator["default"].wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                kwargs = _args.length > 0 && _args[0] !== undefined ? _args[0] : {};
                limit = kwargs === null || kwargs === void 0 ? void 0 : kwargs.limit;
                offset = kwargs === null || kwargs === void 0 ? void 0 : kwargs.offset;
                _this$cleaned_limit_o = this.cleaned_limit_offset(limit, offset), _this$cleaned_limit_o2 = (0, _slicedToArray2["default"])(_this$cleaned_limit_o, 2), cleaned_limit = _this$cleaned_limit_o2[0], cleaner_offset = _this$cleaned_limit_o2[1];
                final_url_obj = new URL(this.list_url);
                final_url_obj.searchParams.append("limit", cleaned_limit);
                final_url_obj.searchParams.append("offset", cleaner_offset);
                final_url_string = final_url_obj.href;
                headers = _objectSpread(_objectSpread({}, this.__headers), this.__dynamic_headers());
                signature = (0, _signature["default"])(final_url_string, "GET", "", headers, this.config.workspace_secret);
                headers["Authorization"] = "".concat(this.config.workspace_key, ":").concat(signature);
                _context.prev = 11;
                _context.next = 14;
                return _axios["default"].get(final_url_string, {
                  headers: headers
                });
              case 14:
                response = _context.sent;
                return _context.abrupt("return", response.data);
              case 18:
                _context.prev = 18;
                _context.t0 = _context["catch"](11);
                throw new _utils.SuprsendApiError(_context.t0);
              case 21:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this, [[11, 18]]);
      }));
      function list() {
        return _list.apply(this, arguments);
      }
      return list;
    }()
  }, {
    key: "detail_url",
    value: function detail_url() {
      var brand_id = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : "";
      var cleaned_brand_id = brand_id.toString().trim();
      var brand_id_encoded = encodeURI(cleaned_brand_id);
      var url = "".concat(this.list_url).concat(brand_id_encoded, "/");
      return url;
    }
  }, {
    key: "get",
    value: function () {
      var _get = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee2() {
        var brand_id,
          url,
          headers,
          signature,
          response,
          _args2 = arguments;
        return _regenerator["default"].wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                brand_id = _args2.length > 0 && _args2[0] !== undefined ? _args2[0] : "";
                url = this.detail_url(brand_id);
                headers = _objectSpread(_objectSpread({}, this.__headers), this.__dynamic_headers());
                signature = (0, _signature["default"])(url, "GET", "", headers, this.config.workspace_secret);
                headers["Authorization"] = "".concat(this.config.workspace_key, ":").concat(signature);
                _context2.prev = 5;
                _context2.next = 8;
                return _axios["default"].get(url, {
                  headers: headers
                });
              case 8:
                response = _context2.sent;
                return _context2.abrupt("return", response.data);
              case 12:
                _context2.prev = 12;
                _context2.t0 = _context2["catch"](5);
                throw new _utils.SuprsendApiError(_context2.t0);
              case 15:
              case "end":
                return _context2.stop();
            }
          }
        }, _callee2, this, [[5, 12]]);
      }));
      function get() {
        return _get.apply(this, arguments);
      }
      return get;
    }()
  }, {
    key: "upsert",
    value: function () {
      var _upsert = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee3() {
        var brand_id,
          brand_payload,
          url,
          headers,
          content_text,
          signature,
          response,
          _args3 = arguments;
        return _regenerator["default"].wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                brand_id = _args3.length > 0 && _args3[0] !== undefined ? _args3[0] : "";
                brand_payload = _args3.length > 1 && _args3[1] !== undefined ? _args3[1] : {};
                url = this.detail_url(brand_id);
                headers = _objectSpread(_objectSpread({}, this.__headers), this.__dynamic_headers());
                content_text = JSON.stringify(brand_payload);
                signature = (0, _signature["default"])(url, "POST", content_text, headers, this.config.workspace_secret);
                headers["Authorization"] = "".concat(this.config.workspace_key, ":").concat(signature);
                _context3.prev = 7;
                _context3.next = 10;
                return _axios["default"].post(url, content_text, {
                  headers: headers
                });
              case 10:
                response = _context3.sent;
                return _context3.abrupt("return", response.data);
              case 14:
                _context3.prev = 14;
                _context3.t0 = _context3["catch"](7);
                throw new _utils.SuprsendApiError(_context3.t0);
              case 17:
              case "end":
                return _context3.stop();
            }
          }
        }, _callee3, this, [[7, 14]]);
      }));
      function upsert() {
        return _upsert.apply(this, arguments);
      }
      return upsert;
    }()
  }]);
  return BrandsApi;
}();
var _default = BrandsApi;
exports["default"] = _default;
module.exports = exports.default;
